### PR TITLE
Add support to display images in nodes

### PR DESCRIPTION
**Pull request by godiard**
Mind maps are useful tools in education. One feature usually requested when are used by kids is the support of images in nodes. 

This pull request add support to add nodes where a image is displayed, without text. Load/save and screenshots are implemented. 

The image can be noted by a url or the data url. In the case od a image noted by a url pointing to a local resource, like the example implemented, a screenshot is not possible, because the canvas is tainted. When the image is selected by the user (or stored as data in the json file) there are no problems.

### CODE REVIEW

File: js/jsmind.screenshot.js [hunk 1]
```diff
@@ -103,19 +112,23 @@
...
-            this._watermark();
-            var url = this.canvas_elem.toDataURL();
-            this.clean();
-            return url;
+            var jsm = this;
+            this._draw(function() {
+                this._watermark();
+                var url = jsm.canvas_elem.toDataURL();
+                jsm.clean();
+                return url;
```
Comment (hizzgdev) [inline js/jsmind.screenshot.js]: this `return` may be wrong.

File: js/jsmind.screenshot.js [hunk 2]
```diff
@@ -197,13 +227,23 @@
...
+                node.ready = true;
+            }
+            if (!!node.image) {
+                var backgroundUrl = css(ncs,'background-image').slice(5, -2);
+                jcanvas.image(ctx, backgroundUrl, tb.x, tb.y, tb.w, tb.h,
+                    function() {
+                        node.ready = true;
+                    });
+                node.ready = false;
             }
```
Comment (hizzgdev) [inline js/jsmind.screenshot.js]: what about if `node.topic` and `node.image` are both exists?

**Comment by hizzgdev** [comment]
great idea and wonderful work.

i will merge it after full testing. thank you very much.

File: js/jsmind.screenshot.js (hunk 1 above)
Comment (godiard) [inline js/jsmind.screenshot.js]: True. Probably the only way to return this value is using a callback function. What you think?
Also, I didn't found a use of shootAsDataURL()

File: js/jsmind.screenshot.js (hunk 2 above)
Comment (godiard) [inline js/jsmind.screenshot.js]: Text will be displayed over the image. Do you think we should make it different?

**Comment by godiard** [comment]
Thanks. I wait for your comments and will do the improvements needed.

Different question: do ou plan merge https://github.com/hizzgdev/jsmind/pull/39 ?

File: js/jsmind.screenshot.js (hunk 2 above)
Comment (ttsdzzg) [inline js/jsmind.screenshot.js]: first `jcanvas.text_ellipsis(...)` then `jcanvas.image(...)`, i think the text will be covered by the image, but without any verification, i will try later :)

File: js/jsmind.screenshot.js (hunk 1 above)
Comment (ttsdzzg) [inline js/jsmind.screenshot.js]: let me think about it, there must be a way out

**Comment by ttsdzzg** [comment]
yes, i will merge both https://github.com/hizzgdev/jsmind/pull/39 and https://github.com/hizzgdev/jsmind/pull/42

File: js/jsmind.screenshot.js (hunk 1 above)
Comment (hizzgdev) [inline js/jsmind.screenshot.js]: let me think about it, there must be a way out

File: js/jsmind.screenshot.js (hunk 2 above)
Comment (hizzgdev) [inline js/jsmind.screenshot.js]: first jcanvas.text_ellipsis(...) then jcanvas.image(...), i think the text will be covered by the image, but without any verification, i will try later :)

**Comment by hizzgdev** [comment]
yes, i will merge both https://github.com/hizzgdev/jsmind/pull/39 and https://github.com/hizzgdev/jsmind/pull/42

File: js/jsmind.screenshot.js (hunk 1 above)
Comment (godiard) [inline js/jsmind.screenshot.js]: What you think about:
```
        shootAsDataURL:function(callback){
            this.init();
            this._watermark();
            var jsm = this;
            this._draw(function() {
                var url = jsm.canvas_elem.toDataURL();
                jsm.clean();
                callback(url);
            });
        },
```

File: js/jsmind.screenshot.js (hunk 2 above)
Comment (godiard) [inline js/jsmind.screenshot.js]: Ok, true. In the case of the screenshot, we should draw first the image and later the text.

**Comment by godiard** [comment]
Tell me when you finish the review, and I will send a new pr to replace #42 with the fixes

**Comment by hizzgdev** [comment]
https://github.com/hizzgdev/jsmind/pull/39 merged

File: js/jsmind.js [hunk 3]
```diff
@@ -2096,13 +2104,41 @@
...
+                if (node.image.startsWith('data')) {
+                    var img = new Image();
+                    img.onload = function() {
+                        var c = document.createElement('canvas');
+                        c.width = imageSize;
+                        c.height = imageSize;
+                        var img = this;
+                        if(c.getContext) {
+                            var ctx = c.getContext('2d');
+                            ctx.drawImage(img, 2, 2, imageSize, imageSize);
```
Comment (hizzgdev) [inline js/jsmind.js]: i don't think fixed size is good, but i have no idea about how to improve it :(

Comment (godiard) [inline js/jsmind.js]: The user can change the default image in the options, or we can add logic to enlarge the image later. In the last case, we will need preserve the original image if we don't want get pixelated images. In my use case, I am focused in simple use for kids.

Comment (godiard) [inline js/jsmind.js]: s/default image/default image size/

**Pull request closed without merging**